File: src/com/chbachman/cminus/representation/statement/FunctionCall.java
package com.chbachman.cminus.representation.statement;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.Struct;
import com.chbachman.cminus.representation.function.Function;
import com.chbachman.cminus.representation.Scope;
import com.chbachman.cminus.representation.Type;
import com.chbachman.cminus.representation.value.Value;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Created by Chandler on 4/12/17.
 */
public class FunctionCall implements Value, Statement {

    Function ref;
    List<Value> parameters;

    public FunctionCall(CMinusParser.FunctionCallContext ctx, Scope scope) {
        this(ctx, scope, makeFunction(ctx, scope));
    }

    protected FunctionCall(CMinusParser.FunctionCallContext ctx, Scope scope, Function ref) {
        this.ref = ref;
        String name = ctx.ID().getText();
        parameters = ctx.value().stream().map(p -> Value.parse(p, scope)).collect(Collectors.toList());

        if (parameters.size() != ref.parameters.size()) {
            throw new RuntimeException("Function " + name + " does not have " + parameters.size() + " parameters.");
        }

        for (int i = 0; i < parameters.size(); i++) {
            if (parameters.get(i).type() != ref.parameters.get(i).type()) {
                throw new RuntimeException("Function " + name + " does not have "
                        + parameters.get(i).type().code() + " as a parameter");
            }
        }
    }

    private static Function makeFunction(CMinusParser.FunctionCallContext ctx, Scope scope) {
        String name = ctx.ID().getText();
        List<Value> parameters = ctx.value().stream().map(p -> Value.parse(p, scope)).collect(Collectors.toList());
        Optional<Function> func = scope.getFunction(name, parameters);
        Function ref;

        if (func.isPresent()) {
            ref = func.get();
        } else {
            throw new RuntimeException("Function " + name + " was not found.");
        }

        return ref;
    }

    @Override
    public Type type() {
        return ref.type();
    }

    @Override
    public String code() {
        return value() + ';';
    }

    @Override
    public String value() {
        StringBuilder s = new StringBuilder(ref.getCName())
                .append('(');

        for (Value value: parameters) {
            s.append(value.value()).append(", ");
        }

        if (!parameters.isEmpty()) {
            s.delete(s.length() - 2, s.length());
        }

        s.append(')');

        return s.toString();
    }
}

File: src/main/java/com/chbachman/cminus/representation/statement/FunctionCall.kt
package com.chbachman.cminus.representation.statement

import com.chbachman.cminus.gen.CMinusParser
import com.chbachman.cminus.representation.Scope
import com.chbachman.cminus.representation.Type
import com.chbachman.cminus.representation.function.Header
import com.chbachman.cminus.representation.value.Expression
import com.chbachman.cminus.representation.value.Identifier
import com.chbachman.cminus.representation.value.Statement
import com.chbachman.cminus.representation.get

/**
 * Created by Chandler on 4/12/17.
 * Represents a function call.
 */

interface Call: Expression, Statement {
    val ref: Header
    val parameters: List<Expression>
    val name: Identifier

    fun verify() {
        if (parameters.size != ref.parameters.size) {
            throw RuntimeException("Function $name does not have ${parameters.size} parameters.")
        }

        for (i in 0..parameters.size - 1) {
            if (parameters[i].type !== ref.parameters[i].type) {
                throw RuntimeException("Function $name does not have ${parameters[i].type.code()} as a parameter")
            }
        }
    }

    override val type: Type
        get() = ref.type

    override val statement: String
        get() = expression + ';'

    override val expression: String
        get() {
            val s = StringBuilder(ref.cName)
                    .append('(')

            for (value in parameters) {
                s.append(value.expression).append(", ")
            }

            if (!parameters.isEmpty()) {
                s.delete(s.length - 2, s.length)
            }

            s.append(')')

            return s.toString()
        }
}

class FunctionCall(ctx: CMinusParser.FunctionCallContext, scope: Scope): Call {
    override val name = Identifier(ctx, scope)
    override val parameters: List<Expression>
    override val ref: Header

    init {
        val func = name.last as Identifier.FuncSegment
        parameters = func.parameters
        ref = func.ref
        verify()
    }
}